### DISCUSSION TITLE
"Error: @prisma/client did not initialize yet" after updating to @keystone-6/core@5.0.0

### DISCUSSION THREAD

**Discussion by DiesIrae** [keystonejs/keystone#8359] [Questions] [answered]
It worked fine in `@keystone-6/core@4.0.1`

publish.Dockerfile
```
FROM node:16.13.0

ENV DATABASE_URL=""
ENV SESSION_SECRET=""

RUN mkdir -p /app && chown -R node:node /app
RUN apt update -y && apt install git

WORKDIR /app

COPY . .

RUN yarn ci
RUN yarn build

EXPOSE 80

ENTRYPOINT ["bash", "runCms.sh"]
CMD ["start"]
```

runCms.sh
```
#!/usr/bin/env bash

set -e

yarn ks:migrateDb

yarn "${@}"
```

package.json
```json
{
  ...
  "scripts": {
    "start": "keystone start",
    "build": "keystone build && yarn cache clean",
    "postinstall": "[ -z ${NO_POST_INSTALL+x} ] && echo 'postinstall' && yarn run ks:postinstall || echo 'no postinstall'",
    "ks:postinstall": "keystone postinstall",
    "ks:migrateDb": "keystone prisma migrate deploy",
    "ci": "rm -rf node_modules && yarn install --frozen-lockfile"
  },
  "dependencies": {
    "@keystone-6/auth": "^7.0.0",
    "@keystone-6/cloudinary": "^7.0.0",
    "@keystone-6/core": "^5.0.0",
    "@keystone-6/fields-document": "^7.0.0",
    ...
   }
}
```

When running the docker image:
```
$ keystone start
✨ Starting Keystone
Error: @prisma/client did not initialize yet. Please run "prisma generate" and try to import it again.
In case this error is unexpected for you, please report it in https://github.com/prisma/prisma/issues
    at new PrismaClient (/app/node_modules/@keystone-6/core/node_modules/.prisma/client/index.js:3:11)
    at getKeystone (/app/node_modules/@keystone-6/core/dist/createSystem-a24c40c1.cjs.dev.js:2154:28)
    at start (/app/node_modules/@keystone-6/core/scripts/cli/dist/keystone-6-core-scripts-cli.cjs.dev.js:1226:20)
    at Object.cli (/app/node_modules/@keystone-6/core/scripts/cli/dist/keystone-6-core-scripts-cli.cjs.dev.js:1369:12)
    at Object.<anonymous> (/app/node_modules/@keystone-6/core/scripts/dist/keystone-6-core-scripts.cjs.dev.js:78:42)
    at Module._compile (node:internal/modules/cjs/loader:1101:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1153:10)
    at Module.load (node:internal/modules/cjs/loader:981:32)
    at Function.Module._load (node:internal/modules/cjs/loader:822:12)
    at Module.require (node:internal/modules/cjs/loader:1005:19)
```

Putting `prismaClientPath: "node_modules/.prisma/client"` in config.db fixed the problem! It re-introduced the line `output   = "node_modules/.prisma/client"` in schema.prisma (commited) that was removed by the Keystone update.
But I don't understand why I need to set prismaClientPath, my Keystone is used as standard, not in a Next app.

Seems linked with [that PR](https://github.com/keystonejs/keystone/pull/8307) but it should have fall back same default so maybe a bug?

More generally, should `output` be set in schema.prisma?
If yes, then the problem is in the development environment, when running postinstall maybe?
If no, then the problem must be in the docker environment.

Docker yarn version: v1.22.15
Dev yarn version: v1.22.19 (the one where schema.prisma is generated)
Dev node version: v19.6.0

**Comment by dcousens** [comment]
@DiesIrae maybe try configuring a custom `db.prismaClientPath` that isn't in your `node_modules` (the default is `node_modules/.prisma/client`).

(see https://github.com/keystonejs/keystone/pull/8307, https://github.com/keystonejs/keystone/pull/8323)

**Reply by DiesIrae** [reply]
Hi @dcousens,

Thanks for your answer. As I wrote in my issue, adding db.prismaClientPath indeed fixed the problem. But I didn't add a custom one, I added the default one, which shouldn't be needed.

**Reply by dcousens** [reply]
@DiesIrae if setting `db.prismaClientPath` to `node_modules/.prisma/client` fixed your problem, then somehow `@prisma/client` (with an `@` at the beginning) is not resolving `.prisma/client` to the `node_modules` you expect.

Why it might not be resolving, probably comes down to how your project is setup.

**Comment by borisno2** [comment]
Hey @DiesIrae I have done a little bit of testing on this and using the following (slightly stripped back) `Dockerfile` and leaving out `db.prismaClientPath` works OK for me
```docker
FROM node:16.13.0

ENV DATABASE_URL=""
ENV SESSION_SECRET=""

RUN mkdir -p /app && chown -R node:node /app
RUN apt update -y && apt install git

WORKDIR /app

COPY . .

RUN yarn ci
RUN yarn build

EXPOSE 80

CMD ["yarn", "start"]
```

With the following scripts in my `package.json`
```
"scripts": {
	"dev": "keystone dev",
	"start": "keystone start",
	"build": "keystone build",
	"postinstall": "keystone postinstall",
	"ci": "rm -rf node_modules && yarn install --frozen-lockfile"
},
```

> More generally, should output be set in schema.prisma?

If you don't set `db.prismaClientPath` the output line should not be printed in your `schema.prisma`

**Reply by dcousens** [reply]
@borisno2 was `yarn cache clean` affecting the `node_modules`?

**Reply by borisno2** [reply]
I tried adding `yarn cache clean` after the build and it still works, also tried adding a script similar to @DiesIrae and it all still works for me without needing to add the `db.prismaClientPath`

I am running on the `5.1.0` release of `@keystone-6/core` so maybe it is something that was resolved with the bump to Prisma 4.11?

**Reply by DiesIrae** [reply]
I updated to 5.1.0 and still have the problem.

And I realize I now also have it when running keystone dev:
```
/usr/local/bin/yarn run dev
yarn run v1.22.19
$ keystone dev
✨ Starting Keystone
⭐️ Server listening on :3000 (http://localhost:3000/)
⭐️ GraphQL API available at /api/graphql
✨ Generating GraphQL and Prisma schemas
✨ Your database is up to date, no migrations need to be created or applied
Error: @prisma/client did not initialize yet. Please run "prisma generate" and try to import it again.
In case this error is unexpected for you, please report it in https://github.com/prisma/prisma/issues
    at new PrismaClient (/Users/richertb/Documents/workspace/MatchaCMS/node_modules/@keystone-6/core/node_modules/.prisma/client/index.js:3:11)
    at getKeystone (/Users/richertb/Documents/workspace/MatchaCMS/node_modules/@keystone-6/core/dist/createSystem-3c7345de.cjs.dev.js:2152:28)
    at setupInitialKeystone (/Users/richertb/Documents/workspace/MatchaCMS/node_modules/@keystone-6/core/scripts/cli/dist/keystone-6-core-scripts-cli.cjs.dev.js:1125:22)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async initKeystone (/Users/richertb/Documents/workspace/MatchaCMS/node_modules/@keystone-6/core/scripts/cli/dist/keystone-6-core-scripts-cli.cjs.dev.js:905:35)
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.

Process finished with exit code 1

```
I tried to remove all my node modules and make a fresh install, still didn't work without forcing  `db.prismaClientPath`. 😢

**Comment by DiesIrae** [comment, marked as answer]
@dcousens @borisno2 thank you both for your answers, that lead me to investigate more and find the root of the problem:

I had `"prisma": "^4.11.0",` in my devDependencies. I use it sometimes for imports. Removing it seemed to fix the problem! I thought as long as the version is the same as Keystone's, it couldn't be a problem.

**Reply by dcousens** [reply]
> I thought as long as the version is the same as Keystone's, it couldn't be a problem.

It probably isn't a problem, but it might result in your `@prisma/client` resolving `.prisma/client` to a different location than you (or prisma) might expect, depending on how your package manager is resolving dependencies.

**Reply by DiesIrae** [reply]
Weird, but removing prisma fixed the problem in dev, but didn't fix the problem in our docker servers. I came back to adding manually `prismaClientPath: "node_modules/.prisma/client"` to db. 

I don't have time to investigate further right now, as it works with this workaround. Apart if you have other ideas.
